cloud/packet/public/resources/device.go

```diff
@@ -0,0 +1,368 @@
+// Copyright © 2017 The Kubicorn Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/kris-nova/kubicorn/apis/cluster"
+	"github.com/kris-nova/kubicorn/cloud"
+	"github.com/kris-nova/kubicorn/cutil/compare"
+	"github.com/kris-nova/kubicorn/cutil/defaults"
+	"github.com/kris-nova/kubicorn/cutil/logger"
+	"github.com/kris-nova/kubicorn/cutil/script"
+	"github.com/packethost/packngo"
+)
+
+const (
+	MasterIPAttempts               = 100
+	MasterIPSleepSecondsPerAttempt = 5
+)
+
+var _ cloud.Resource = &Device{}
+
+type Device struct {
+	Shared
+	Location         string
+	Type             string
+	OS               string
+	SSHFingerprint   string
+	BootstrapScripts []string
+	Count            int
+	ServerPool       *cluster.ServerPool
+	Tags             []string
+	ProjectID        string
+}
+
+func (r *Device) Actual(immutable *cluster.Cluster) (*cluster.Cluster, cloud.Resource, error) {
+	logger.Debug("device.Actual")
+	newResource := &Device{
+		Shared: Shared{
+			Name: r.Name,
+		},
+	}
+
+	var valid []packngo.Device
+	logger.Debug("device.Actual finding project ID by name %s", immutable.Project.Name)
+	project, err := GetProjectByName(immutable.Project.Name)
+	if err != nil {
+		return nil, nil, err
+	}
+	if project != nil && project.ID != "" {
+		tag := r.ServerPool.Type
+		logger.Debug("device.Actual getting devices with tag %s", tag)
+		valid, err = devicesByTag(project.ID, tag)
+		logger.Debug("device.Actual devices found %v", valid)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if len(valid) > 0 {
+		newResource.Count = len(valid)
+		device := valid[0]
+		newResource.Name = r.Name
+		newResource.Type = device.Plan.Slug
+		newResource.OS = device.OS.Slug
+		newResource.Location = device.Facility.Code
+		newResource.BootstrapScripts = r.ServerPool.BootstrapScripts
+		newResource.SSHFingerprint = immutable.SSH.PublicKeyFingerprint
+		newResource.Tags = device.Tags
+		newResource.ProjectID = project.ID
+
+	} else {
+		newResource.Count = 0
+		newResource.Type = r.ServerPool.Size
+		newResource.OS = r.ServerPool.Image
+		newResource.Location = r.Location
+		newResource.BootstrapScripts = r.ServerPool.BootstrapScripts
+		newResource.SSHFingerprint = immutable.SSH.PublicKeyFingerprint
+		newResource.Name = r.ServerPool.Name
+		newResource.Tags = r.Tags
+	}
+
+	fmt.Println(immutable)
```

Comment:
I don't believe that `fmt` is meant to be here.

Reply:
Ugh! One issue I had when working through it, threw that in because of line and timing issues and never deleted. Nice catch...